### DISCUSSION TITLE
Where does `outfile` field in the output structure of in a BinaryFile retrieving API come from ?

### DISCUSSION THREAD

**Discussion by Zhenye-Na** [aws/aws-cli#6960] [Q&A] [answered]
### Background

We utilized [aws cli add-model](https://docs.aws.amazon.com/cli/latest/reference/configure/add-model.html), imported the json spec and configure to enable the cli commands, all works perfectly

### Question

In one API we built, the output of the API is trying to return a binary file, like return a zip archive or a text file like API Gateway [get-export](https://docs.aws.amazon.com/cli/latest/reference/apigateway/get-export.html)

I noticed that when I try to invoke the command with help to see the information, I see

```
aws awesome-service awesome-function help


XXXX
XXXX


outfile (string) Filename where the content will be saved
```

However, I did not define any attribute in my output structure to have this variable name.

Can I get some instructions how and why the `outfile` appeared in my CLI command and anything I can do to customize it?

for example, changed it to 

```
--output-file filename
```

and by default create a `filename.ext` in the local

**Comment by tim-finnigan** [comment, marked as answer]
The outfile argument is going to be a positional argument representing the name of the file to save the binary payload from your operation.

So for example, you would run:

`aws awesome-service awesome-function path/to/payload-content`

and it would save the content of the payload to the file path/to/payload-content.

This argument is added automatically to any CLI command where the API operation has an output that is a streaming blob. In terms of customizing it, the team would prefer to keep what is automatically added to stay consistent with other service APIs whose outputs are streaming blobs.

**Reply by Zhenye-Na** [reply]
Thank you for the response. Exactly what I found right now.

My only concern is that, suppose our backend API return the blob of a `zip file`, but customer assigns `outfile` argument as a text file, or any other file extension, for example

```
aws awesome-service awesome-function plain-text.txt
aws awesome-service awesome-function tar-ball.tar

// plain-text.txt or tar-ball.tar is wrong because it wont be opened properly
```

This will lead to that customer receives incorrect result. so what I am thinking to `customize` is something like

```
aws awesome-service awesome-function --out-file-name awesome-filename
```

and behind the scene, the cli will download the file blob, to `awesome-filename`**`.zip`**,

I see there are some customization in other AWS services [awscli/customizations](https://github.com/aws/aws-cli/tree/develop/awscli/customizations). I am wondering if this will be the right place if I need to modify the CLI behavior ?

**Reply by Zhenye-Na** [reply]
@tim-finnigan 

any insights forthe follow up I asked here?

**Reply by tim-finnigan** [reply]
Since this question was raised internally we asked that the discussion continue in the ticket you originally created.

**Reply by Zhenye-Na** [reply]
found it, thanks!
